### PR TITLE
Brigadier DSL enhancements and type-safe argument readers

### PR DESCRIPTION
**Pull request by Cypher121**
Changes made and progress:

- [x] Clean up typos and misformatted params in KDoc
- [x] Split argument builders into files based on rough categories
- [x] Create a result (success/failure) returning wrapper for `executes(Command<S>)`
- [x] Create a Unit returning wrapper for `executes(Command<S>)` (succeeds with SINGLE_SUCCESS unless exception thrown)
- [x] Implement accessors for QSL's EnumArgumentType
  - [ ] Move accessors to wrappers/qsl (requires /qsl to depend on /minecraft?)
  - [x] Remove the unsafe cast on QSL's enum, requires QuiltMC/quilt-standard-libraries#147
- [x] Safe accessors
  - [x] Implement context value accessors for argument types with known resolution methods for an unspecified command source
  - [x] Implement builders with name as lambda parameter for argument types without known resolution methods for an unspecified command source
  - [x] Implement builder overloads with context value accessors for command source and argument type combos with known resolution methods
- [x] Write missing KDocs
- [x] Provide example usage in KDocs or sample folders?

### CODE REVIEW

File: wrapper/minecraft/src/main/kotlin/org/quiltmc/qkl/wrapper/minecraft/brigadier/ArgumentConstructor.kt [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+    /**
+     * Converts this constructor into a required argument.
+     *
+     * @see CommandArgument.Required
+     * @see required
+     *
+     * @author Cypher121
+     */
+    public fun required(): CommandArgument.Required<S, B, D> =
+        CommandArgument.Required(builder, name, descriptor)
```
Comment (Oliver-makes-code) [inline wrapper/minecraft/src/main/kotlin/org/quiltmc/qkl/wrapper/minecraft/brigadier/ArgumentConstructor.kt:45]: Make sure to avoid `fun =` syntax

**Review by Oliver-makes-code — CHANGES_REQUESTED** [review]
The only real gripe I have is with the use of `fun =` syntax

File: wrapper/minecraft/src/main/kotlin/org/quiltmc/qkl/wrapper/minecraft/brigadier/ArgumentConstructor.kt (hunk 1 above)
Comment (Cypher121) [reply wrapper/minecraft/src/main/kotlin/org/quiltmc/qkl/wrapper/minecraft/brigadier/ArgumentConstructor.kt:45]: what's the reasoning behind that one actually? /gen

Comment (Cypher121) [reply wrapper/minecraft/src/main/kotlin/org/quiltmc/qkl/wrapper/minecraft/brigadier/ArgumentConstructor.kt]: Fix

Comment (Oliver-makes-code) [reply wrapper/minecraft/src/main/kotlin/org/quiltmc/qkl/wrapper/minecraft/brigadier/ArgumentConstructor.kt]: > what's the reasoning behind that one actually? /gen

Consistency

File: wrapper/minecraft/src/main/kotlin/org/quiltmc/qkl/wrapper/minecraft/brigadier/argument/IdentifierArguments.kt [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+ *
+ * @author Cypher121
+ */
+@BrigadierDsl
+public fun ArgumentReader<
+        ServerCommandSource,
+        DefaultArgumentDescriptor<IdentifierArgumentType>
+        >.asAdvancement(): Advancement {
+    return IdentifierArgumentType.getAdvancementArgument(context, name)
+}
```
Comment (Oliver-makes-code) [inline wrapper/minecraft/src/main/kotlin/org/quiltmc/qkl/wrapper/minecraft/brigadier/argument/IdentifierArguments.kt:64]: Could be better optimized for readability

File: wrapper/minecraft/src/main/kotlin/org/quiltmc/qkl/wrapper/minecraft/brigadier/argument/IdentifierArguments.kt [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+ *
+ * @author Cypher121
+ */
+@BrigadierDsl
+public fun ArgumentReader<
+        ServerCommandSource,
+        DefaultArgumentDescriptor<IdentifierArgumentType>
+        >.asPredicate(): LootCondition {
+    return IdentifierArgumentType.getPredicateArgument(context, name)
+}
```
Comment (Oliver-makes-code) [inline wrapper/minecraft/src/main/kotlin/org/quiltmc/qkl/wrapper/minecraft/brigadier/argument/IdentifierArguments.kt:81]: This is the last one I'm gonna put the "Could be better optimized for readability" on, but make sure everything's readable

File: wrapper/minecraft/src/main/kotlin/org/quiltmc/qkl/wrapper/minecraft/brigadier/argument/MiscArguments.kt [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+
+/**
+ * Descriptor for a literal argument.
+ *
+ * Separate from [DefaultArgumentDescriptor]
+ * to stand out more in type hints.
+ *
+ * @author Cypher121
+ */
+public object LiteralDescriptor : ArgumentDescriptor<ArgumentType<*>>
```
Comment (Oliver-makes-code) [inline wrapper/minecraft/src/main/kotlin/org/quiltmc/qkl/wrapper/minecraft/brigadier/argument/MiscArguments.kt:112]: Make sure to keep at the top of the file

File: wrapper/minecraft/src/main/kotlin/org/quiltmc/qkl/wrapper/minecraft/brigadier/argument/PlayerArguments.kt [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+/**
+ * [ArgumentDescriptor] for an [EntityArgumentType]
+ * allowing multiple players to be selected.
+ *
+ * @see players
+ * @see EntityArgumentType.players
+ *
+ * @author Cypher121
+ */
+public object ListPlayerArgumentDescriptor : ArgumentDescriptor<EntityArgumentType>
```
Comment (Oliver-makes-code) [inline wrapper/minecraft/src/main/kotlin/org/quiltmc/qkl/wrapper/minecraft/brigadier/argument/PlayerArguments.kt:186]: Make sure to keep at the top of the file

File: wrapper/minecraft/src/main/kotlin/org/quiltmc/qkl/wrapper/minecraft/brigadier/argument/ScoreboardArguments.kt [hunk 6]
```diff
@@ -0,0 +1,260 @@
...
+/**
+ * [ArgumentDescriptor] for an [ScoreHolderArgumentType]
+ * allowing multiple score holders to be selected.
+ *
+ * @see scoreHolders
+ * @see ScoreHolderArgumentType.scoreHolders
+ *
+ * @author Cypher121
+ */
+public object ListScoreHolderArgumentDescriptor : ArgumentDescriptor<ScoreHolderArgumentType>
```
Comment (Oliver-makes-code) [inline wrapper/minecraft/src/main/kotlin/org/quiltmc/qkl/wrapper/minecraft/brigadier/argument/ScoreboardArguments.kt:260]: Make sure to keep at the top of the file

**Review by Oliver-makes-code — CHANGES_REQUESTED** [review]

File: wrapper/minecraft/src/main/kotlin/org/quiltmc/qkl/wrapper/minecraft/brigadier/util/ServerCommandUtil.kt [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+ *
+ * @see ServerPlayerEntity.sendMessage
+ *
+ * @author Cypher121
+ */
+public fun CommandContext<ServerCommandSource>.respond(
+    message: Text,
+    messageType: RegistryKey<MessageType> = MessageType.SYSTEM
+) {
+    player?.sendMessage(message, messageType)
```
Comment (SilverAndro) [inline wrapper/minecraft/src/main/kotlin/org/quiltmc/qkl/wrapper/minecraft/brigadier/util/ServerCommandUtil.kt:84]: This should probably be `source.sendFeedback` (iirc thats the name?)

**Review by SilverAndro — CHANGES_REQUESTED** [review]

File: wrapper/minecraft/src/main/kotlin/org/quiltmc/qkl/wrapper/minecraft/brigadier/util/ServerCommandUtil.kt (hunk 7 above)
Comment (Cypher121) [reply wrapper/minecraft/src/main/kotlin/org/quiltmc/qkl/wrapper/minecraft/brigadier/util/ServerCommandUtil.kt:84]: I may be blind

Comment (Cypher121) [reply wrapper/minecraft/src/main/kotlin/org/quiltmc/qkl/wrapper/minecraft/brigadier/util/ServerCommandUtil.kt]: replaced

**Review by SilverAndro — APPROVED** [review]
Quite the PR! 🥳

**Review by PotatoPresident — APPROVED** [review]
This is awesome, only scares me a little.

**Review by Oliver-makes-code — APPROVED** [review]
looks good

**Pull request merged**